### PR TITLE
Codespaces - Oryx version bump to exp-20210721.2

### PR DESCRIPTION
**Pull request by samruddhikhandale**
Oryx tag - https://github.com/microsoft/Oryx/releases/tag/exp-20210721.2

It contains changes needed for [oryx build status should show the actual command oryx invokes](https://github.com/microsoft/vssaas-planning/issues/3931)

### CODE REVIEW

**Comment by jkeech** [comment]
@samruddhikhandale, is this "exp" tag ready for production use?

**Comment by samruddhikhandale** [comment]
@arroyc would you like to confirm if it's ready to be used by the codespace service for production?

**Comment by arroyc** [comment]
> @arroyc would you like to confirm if it's ready to be used by the codespace service for production?

yes it is ..

**Comment by Chuxel** [comment]
@jkeech Do we want to consider https://github.com/microsoft/vscode-dev-containers/issues/952 for this image version as well? Looks like an update to common debian to make sure no git editor is set before setting GIT_EDITOR (e.g. `if [ -z "${GIT_EDITOR}" ] && [  -z "$(git config --get core.editor)" ]`)

**Comment by jkeech** [comment]
> @jkeech Do we want to consider #952 for this image version as well? Looks like an update to common debian to make sure no git editor is set before setting GIT_EDITOR (e.g. `if [ -z "${GIT_EDITOR}" ] && [ -z "$(git config --get core.editor)" ]; then export GIT_EDITOR="code --wait"; fi`)

I think @joshspicer can pick that up next week when he's back, but I wouldn't hold the current release for it. We can always publish a new version soon after once that's ready.

**Comment by Chuxel** [comment]
@jkeech @arroyc - We're seeing an odd `rvm` error with this image.  I vaguely remember Josh mentioning it.  Do we know if he got it working?  Right now this fails build - I'm not sure if there's an adjustment that he made that is in a branch somewhere.

**Comment by Chuxel** [comment]
@arroyc @samruddhikhandale - Okay, couple problems with the image:

1. The default shell for the `codespace` user is `/bin/sh` instead of `/bin/bash` (pass `-s /bin/bash` into `useradd` to change it). This is what caused the rvm errors (since it works in bash, but not sh).
2. While the privileges of folders under `$HOME` like `~/.dotnet` are owned by codespace, they're symlinks.  The problem is symlinks get the permissions of the folder it points to, which is the folders under `/opt`. Those folders would need to have the owners changed for that to work.

**Comment by Chuxel** [comment]
Oh one more - it looks like ORYXPATHS is set to `ORYX_PATHS=/opt/oryx:/opt/nodejs/lts/bin:/opt/dotnet/lts:/opt/python/latest/bin:/opt/php/lts/bin:/opt/php-composer:/opt/yarn/stable/bin:/opt/hugo/lts::/opt/java/lts/bin:/opt/maven/lts/bin:/opt/ruby/lts/bin`.  Should we be updating the PATH to include locations under /home/codespaces instead or is the ORYX_PATHS value wrong?  (Right now we just add this into the PATH).

**Review by Chuxel — COMMENTED** [review]
As discussed with @arroyc, we need to update the `PATH` to get the dynamic platform stuff to kick in.  I believe this should do it on line 28:

```Dockerfile
ENV PATH="${ORIGINAL_PATH}:${NVM_DIR}/current/bin:${NPM_GLOBAL}/bin:${DOTNET_ROOT}:${DOTNET_ROOT}/tools:${SDKMAN_DIR}/bin:${SDKMAN_DIR}/candidates/gradle/current/bin:${SDKMAN_DIR}/candidates/java/current/bin:${CARGO_HOME}/bin:${GOROOT}/bin:${GOPATH}/bin:${PIPX_BIN_DIR}:/opt/conda/condabin:/home/codespace/.nodejs/current/bin:/home/codespace/current/.java/bin:/home/codespace/.maven/current/bin:/home/codespace/.python/current/bin:/home/codespace/.php/current/bin:/home/codespace/.ruby/current/bin:/home/codespace/.hugo/current/bin:${ORYX_PATHS}"
```
The ORYX_PATHs part then includes the `/opt` folders at the end of the path in case something isn't quite right with the dynamic setup and to catch things like Hugo and Maven that aren't dynamic yet.

One other nit below.

File: containers/codespaces-linux/.devcontainer/base.Dockerfile [hunk 1]
```diff
@@ -59,6 +59,9 @@ RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
     && mkdir -p /usr/local/etc/vscode-dev-containers/ \
     && mv -f /tmp/scripts/first-run-notice.txt /usr/local/etc/vscode-dev-containers/
 
+# Changing default shell (/bin/sh) to /bin/bash
+RUN chsh -s /bin/bash ${USERNAME}
```
[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply containers/codespaces-linux/.devcontainer/base.Dockerfile]: @Chuxel Looks like you missed looking at my last commit https://github.com/microsoft/vscode-dev-containers/pull/958/commits/0580a3be42a230301468a27e89cd1cacc2be5093

It contains the /home/codespace/.<platform> part, which I copied from https://github.com/microsoft/vscode-dev-containers/pull/943/files

I will add the missing parts like hugo and maven now.

[the remark this answers is not in the document]
Comment (Chuxel) [reply containers/codespaces-linux/.devcontainer/base.Dockerfile]: Ha - you are right. The build took so long that I was old by the time it was done. lol

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply containers/codespaces-linux/.devcontainer/base.Dockerfile]: https://github.com/microsoft/vscode-dev-containers/pull/958/commits/622e76c556349e570ab4851be8d94b07029bd790 Added hugo and maven. Also, moved `&& chsh -s /bin/bash ${USERNAME} \` according to your suggestion.
Let me know if this looks good. @Chuxel 😄 
Creating codespace and remote container was a success. 
Smoke test passed as well 💯

**Comment by Chuxel** [comment]
LGTM!  Let's move it into dev.

**Comment by Chuxel** [comment]
Here's the build run that will create `mcr.microsoft.com/vscode/devcontainers/universal:dev-linux` when it completes: https://github.com/microsoft/vscode-dev-containers/actions/runs/1057658307

**Comment by samruddhikhandale** [comment]
Thanks! Once created, I will validate it using the tests mentioned in this pipeline -https://dev.azure.com/devdiv/OnlineServices/_build/results?buildId=4987216&view=logs&j=c75d4970-da9f-5df4-8cfc-498fb498b424&t=cd1e8c0a-ec24-5349-2e99-0600b15d276f&l=53

**Comment by samruddhikhandale** [comment]
`mcr.microsoft.com/vscode/devcontainers/universal:dev-linux` looks good 
(Validated for platforms, shells and docker-in-docker)

**Pull request merged**